examples/training/matryoshka/README.md

```diff
@@ -0,0 +1,54 @@
+# Matryoshka Embeddings
+
+Dense embedding models normally produce embeddings with a fixed size, such as 768 or 1024. All further computations (clustering, classification, semantic search, retrieval, reranking, etc.) must then be done on these full embeddings. [Matryoshka Representation Learning](https://arxiv.org/abs/2205.13147) revisits this idea, and proposes a solution to train embedding models whose embeddings are still useful after truncation to much smaller sizes. This allows for considerably faster (bulk) processing.
+
+## Use Cases
+
+A particularly interesting use cases are to split up processing into two steps: 1) pre-processing with much smaller vectors and then 2) processing the remaining vectors as full size (also called "shortlisting and reranking"). Additionally, Matryoshka models will allow you to scale your embedding solutions to your desired storage cost, processing speed and performance.
+
+## Training
+
+Training using Matryoshka Representation Learning (MRL) is quite elementary: rather than applying some loss function on only the full-size embeddings, we also apply that same loss function on truncated portions of the embeddings. For example, if a model has an embedding dimension of 768 by default, it can now be trained on 768, 512, 256, 128, 64 and 32. Each of these losses will be added together, optionally with some weight:
+
+```python
+from sentence_transformers import SentenceTransformer
+from sentence_transformers.losses import CoSENTLoss, MatryoshkaLoss
+
+model = SentenceTransformer("microsoft/mpnet-base")
+
+base_loss = CoSENTLoss(model=model)
+loss = MatryoshkaLoss(model=model, loss=base_loss, matryoshka_dims=[768, 512, 256, 128, 64])
+```
+* **Reference**: [`MatryoshkaLoss`](../../../docs/package_reference/losses#matryoshkaloss)
+
+## Inference
+
+When a model has been trained using a Matryoshka loss, then you can run inference with it using [`SentenceTransformers.encode`](../../../docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode). You must then truncate the resulting embeddings, and it is recommended to renormalize the embeddings.
+
+```python
+from sentence_transformers import SentenceTransformer
+from sentence_transformers.util import cos_sim
+
+model = SentenceTransformer("nomic-ai/nomic-embed-text-v1.5", trust_remote_code=True)
+
+matryoshka_dim = 64
+embeddings = model.encode(
+    [
+        "search_query: What is TSNE?",
+        "search_document: t-distributed stochastic neighbor embedding (t-SNE) is a statistical method for visualizing high-dimensional data by giving each datapoint a location in a two or three-dimensional map.",
+        "search_document: Amelia Mary Earhart was an American aviation pioneer and writer.",
+    ]
+)
+embeddings[..., :matryoshka_dim]  # Shrink the embedding dimensions
```

Comment:
It might be good to mention the additional steps (as stated in the [model card](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5#sentence-transformers)):
1. Applying layer normalization after encoding (before truncation)
2. (Optional) Normalizing the truncated embeddings. Luckily since we're performing cosine similarity in this example, it doesn't matter, but just so the user is aware of it in case they're going to store the embedding at this point.

Reply:
(for reference)
```py
import torch.nn.functional as F
from sentence_transformers import SentenceTransformer

matryoshka_dim = 512

model = SentenceTransformer("nomic-ai/nomic-embed-text-v1.5", trust_remote_code=True)
sentences = ['search_query: What is TSNE?', 'search_query: Who is Laurens van der Maaten?']
embeddings = model.encode(sentences, convert_to_tensor=True)
embeddings = F.layer_norm(embeddings, normalized_shape=(embeddings.shape[1],))
embeddings = embeddings[:, :matryoshka_dim]
embeddings = F.normalize(embeddings, p=2, dim=1)
print(embeddings)
```

Reply:
Hmm, the tricky part is that this is a Nomic-specific thing I believe, not just for Matryoshka models.

Reply:
yeah this is a Nomic Specific thing but not necessary in general. If you want to include the Nomic Model, you should use the layernorm

Reply:
Hmm, that makes it a bit tricky. I'd love to show a general example that works with essentially all models.

Reply:
I'm seeing extremely minimal differences in performance when using layer norm vs no layer norm. 
| | 768 | 512 | 256 | 128 | 64 |
| --- | ---- |  --- | ---- | --- | ---- |
| Layer norm | 0.84247 |0.83947 | **0.83743** | **0.83185**|0.82377 |
| No layer norm | 0.84247 | 0.83947| 0.83725| 0.83168|**0.82386** |
| Ratio | 1.00000221 | 0.99999807 |  1.00020698 | 1.00020339 | 0.9998845 |

I would rather not recommend the layer norm in these docs, as they're just for general inference (not just Nomic), so either:
1. I use Nomic v1.5 without Layer Norm (should perform roughly equivalently, but not identically as you can see)
2. I use some other (unknown) model, like https://huggingface.co/tomaarsen/mpnet-base-all-nli-matryoshka

My preference is the first option, but if you think that I shouldn't list Nomic without layer norm, then I'll do the second.